### PR TITLE
Fix bug in `IntersectionSearchSpace`

### PR DESCRIPTION
**Pull request by contramundum53**
## Motivation
`IntersectionSearchSpace` has a bug that `self._cursor` is not updated when there are no running trials. We fix this bug.

## Description of the changes
Fix `self._cursor` such that it is updated to the last trial when there are no running trials.

### CODE REVIEW

**Comment by c-bata** [comment]
@not522 @keisuke-umezawa Could you review this PR?

File: optuna/samplers/_search_space/intersection.py [hunk 1]
```diff
@@ -59,20 +59,20 @@ def calculate(self, study: Study, ordered_dict: bool = False) -> Dict[str, BaseD
             if self._study_id != study._study_id:
                 raise ValueError("`IntersectionSearchSpace` cannot handle multiple studies.")
 
-        states_of_interest = [optuna.trial.TrialState.COMPLETE]
+        states_of_interest = [optuna.trial.TrialState.COMPLETE, optuna.trial.TrialState.RUNNING]
```
Comment (not522) [inline optuna/samplers/_search_space/intersection.py:62]:
```suggestion
        states_of_interest = [optuna.trial.TrialState.COMPLETE, optuna.trial.TrialState.WAITING, optuna.trial.TrialState.RUNNING]
```

File: optuna/samplers/_search_space/intersection.py [hunk 2]
```diff
@@ -59,20 +59,20 @@ def calculate(self, study: Study, ordered_dict: bool = False) -> Dict[str, BaseD
...
 
         if self._include_pruned:
             states_of_interest.append(optuna.trial.TrialState.PRUNED)
 
-        next_cursor = self._cursor
-        for trial in reversed(study.get_trials(deepcopy=False)):
+        trials = study.get_trials(deepcopy=False, states=states_of_interest)
+
+        next_cursor = trials[-1].number if len(trials) > 0 else -1
```
Comment (not522) [inline optuna/samplers/_search_space/intersection.py:69]: I think that the cursor should point to the youngest unfinished trial.
```suggestion
        next_cursor = trials[-1].number + 1 if len(trials) > 0 else -1
```

**Review by not522 — COMMENTED** [review]
Could you add tests for this change?

File: optuna/samplers/_search_space/intersection.py (hunk 2 above)
Comment (contramundum53) [reply optuna/samplers/_search_space/intersection.py]: Thanks, I updated the code.

File: optuna/samplers/_search_space/intersection.py (hunk 1 above)
Comment (contramundum53) [reply optuna/samplers/_search_space/intersection.py]: Thanks!

**Comment by contramundum53** [comment]
> Could you add tests for this change?

Should I add tests for `_cursor`? I thought that would count as implementation detail.

**Comment by not522** [comment]
I see, so this bug only slows the execution down but does not change the behavior.

**Review by not522 — APPROVED** [review]
LGTM

File: optuna/samplers/_search_space/intersection.py [hunk 3]
```diff
@@ -59,20 +59,24 @@ def calculate(self, study: Study, ordered_dict: bool = False) -> Dict[str, BaseD
...
+        ]
 
         if self._include_pruned:
             states_of_interest.append(optuna.trial.TrialState.PRUNED)
 
-        next_cursor = self._cursor
-        for trial in reversed(study.get_trials(deepcopy=False)):
+        trials = study.get_trials(deepcopy=False, states=states_of_interest)
+
+        next_cursor = trials[-1].number + 1 if len(trials) > 0 else -1
```
Comment (keisuke-umezawa) [inline optuna/samplers/_search_space/intersection.py:73]: [question] It relies on that the assumption that `get_trials` returns a list of ordered trials, but is it actually guaranteed in all the storages? e.g. [redis](https://github.com/optuna/optuna/blob/master/optuna/storages/_redis.py#L687-L699), [SQL](https://github.com/optuna/optuna/blob/master/optuna/storages/_rdb/storage.py#L953-L962)

File: optuna/samplers/_search_space/intersection.py [hunk 4]
```diff
@@ -59,20 +59,24 @@ def calculate(self, study: Study, ordered_dict: bool = False) -> Dict[str, BaseD
...
+        for trial in reversed(trials):
             if self._cursor > trial.number:
                 break
 
             if not trial.state.is_finished():
```
Comment (keisuke-umezawa) [inline optuna/samplers/_search_space/intersection.py:78]: [question] This clause is continued in the PR, but not in the master branch. Is it ok?

File: optuna/samplers/_search_space/intersection.py (hunk 3 above)
Comment (keisuke-umezawa) [reply optuna/samplers/_search_space/intersection.py:73]: They are always wrapped by cached storage and trials are guaranteed to be ordered.
https://github.com/optuna/optuna/blob/master/optuna/storages/_cached_storage.py#L344

File: optuna/samplers/_search_space/intersection.py [hunk 5]
```diff
@@ -59,20 +59,24 @@ def calculate(self, study: Study, ordered_dict: bool = False) -> Dict[str, BaseD
...
                 next_cursor = trial.number
-
-            if trial.state not in states_of_interest:
                 continue
```
Comment (keisuke-umezawa) [inline optuna/samplers/_search_space/intersection.py:81]: Could you add `assert trial.state == optuna.trial.TrialState.COMPLETE or trial.state == optuna.trial.TrialState.PRUNED` here, to understand this logic easily? Or, could you add unit tests which guarantee that intersections will not count RUNNING and WAITING?

File: optuna/samplers/_search_space/intersection.py (hunk 1 above)
Comment (keisuke-umezawa) [reply optuna/samplers/_search_space/intersection.py]: Could you add unit tests to catch this error? 

Currently, we do not check that `trials in the state of RUNNING, WAITING are not counted, but later they are counted after completion` in unit tets. 
Only we checked that FAIL and PRUNED are skipped here.
https://github.com/optuna/optuna/blob/master/tests/samplers_tests/search_space_tests/test_intersection.py#L54-L55

File: optuna/samplers/_search_space/intersection.py (hunk 3 above)
Comment (contramundum53) [reply optuna/samplers/_search_space/intersection.py:73]: Yes, we assume that `get_trials` do return the trials in chronological order. However, there is a known issue (#3605) that the returned trials can be unsorted. This issue is to be fixed.

File: optuna/samplers/_search_space/intersection.py (hunk 1 above)
Comment (keisuke-umezawa) [reply optuna/samplers/_search_space/intersection.py]: @contramundum53 Sorry, I added a comment on the resolved conversation. Could you check it
?

Comment (contramundum53) [reply optuna/samplers/_search_space/intersection.py]: @keisuke-umezawa I added a test case for waiting trial. PTAL.

**Review by keisuke-umezawa — APPROVED** [review]
Thank you for adding a unittest to catch the error. LTGM.
I will merge it even though a non-critial ci has not passed.

**Pull request merged**